dynamodb/batch/batch.go

```diff
@@ -0,0 +1,150 @@
+	DeleteRequest RequestType = 1
+	// requestLimit is the number of requests in a BatchWriteItem. While the numeric limit from AWS is 25, there is also a size limit in megabytes.
+	requestLimit = 25
+	// After the sleep reaches this value of seconds (for retrying unprocessed items), we error with a time-out
+	maxTimeout int64 = 10
+)
+
+// Batch takes on the job of tracking BatchWrite state
+type Batch struct {
+	table        string
+	bwInput      dynamodb.BatchWriteItemInput
+	requests     []*dynamodb.WriteRequest
+	requestLimit int   // number of requests per batch
+	maxTimeout   int64 // maximum time to sleep between retries
+	dynamoapi    dynamodbiface.DynamoDBAPI
+	sleepSeconds int64 // Backoff
+}
+
+// Iface presents a test interface for Batch
+type Iface interface {
+	SetTableName(tableName string)
+	SetRequestLimit(newLimit int) error
+	SetMaxTimeout(newTimeout int64) error
+	Add(requestType RequestType, item map[string]*dynamodb.AttributeValue) ([]*dynamodb.ConsumedCapacity, error)
+	Send() ([]*dynamodb.ConsumedCapacity, error)
+}
+
+// NewBatch returns a new object with the given table name, request limit, and dynamo API
+func NewBatch(ddb dynamodbiface.DynamoDBAPI) *Batch {
+	return &Batch{
+		dynamoapi:    ddb,
+		sleepSeconds: 1,
+		requestLimit: requestLimit,
+		maxTimeout:   maxTimeout,
+		bwInput:      basicBatchInput(),
+	}
+}
+
+func basicBatchInput() dynamodb.BatchWriteItemInput {
+	return dynamodb.BatchWriteItemInput{
+		RequestItems:           map[string][]*dynamodb.WriteRequest{},
+		ReturnConsumedCapacity: aws.String("TOTAL"),
+	}
+}
+
+// SetTableName sets the table name
+func (b *Batch) SetTableName(tableName string) {
+	b.table = tableName
+	return
+}
+
+// SetRequestLimit sets the number of items to pile up before sending
+func (b *Batch) SetRequestLimit(newLimit int) error {
+	if newLimit > requestLimit {
+		return fmt.Errorf("requestLimit must be %d or less", requestLimit)
+	}
+	if newLimit < 1 {
+		return fmt.Errorf("requestLimit must be %d or less", requestLimit)
+	}
+	b.requestLimit = newLimit
+	return nil
+}
+
+// SetMaxTimeout sets the maximum sleep allowed for retries
+func (b *Batch) SetMaxTimeout(newTimeout int64) error {
+	if newTimeout > maxTimeout {
+		return fmt.Errorf("too much timeout, should be less than %d", maxTimeout)
+	}
+	if newTimeout < 1 {
+		return fmt.Errorf("not enough time out, should be at least 1")
+	}
+	b.maxTimeout = int64(newTimeout)
+	return nil
+}
+
+// Add an item to the batch, and send the batch if it's full
+func (b *Batch) Add(requestType RequestType, item map[string]*dynamodb.AttributeValue) ([]*dynamodb.ConsumedCapacity, error) {
+	if b.table == "" {
+		return nil, fmt.Errorf("table name required, call SetTableName")
+	}
+	request := &dynamodb.WriteRequest{}
+	switch requestType {
+	case PutRequest:
+		request.PutRequest = &dynamodb.PutRequest{Item: item}
+	case DeleteRequest:
+		request.DeleteRequest = &dynamodb.DeleteRequest{Key: item}
+	}
```

Comment:
`requestType` is baffling me. First, why not use an iota as an enum if going this route? You could also do it how the http package does methods, where they alias strings rather than numbers, so either work. But why not just have a method that is explicit, `AddPut` / `AddDelete`, then have private `add` take a `WriteRequest`? This works as is no doubt, but it seems awkward. This will not block, but I am curious.